src/cluster_legacy.c

```diff
@@ -3312,10 +3312,33 @@ int clusterProcessPacket(clusterLink *link) {
                     if (sender->replicaof) clusterNodeRemoveReplica(sender->replicaof, sender);
                     serverLog(LL_NOTICE, "Node %.40s (%s) is now a replica of node %.40s (%s) in shard %.40s",
                               sender->name, sender->human_nodename, sender_claimed_primary->name,
-                              sender_claimed_primary->human_nodename, sender->shard_id);
+                              sender_claimed_primary->human_nodename, sender_claimed_primary->shard_id);
                     clusterNodeAddReplica(sender_claimed_primary, sender);
                     sender->replicaof = sender_claimed_primary;
 
+                    /* Currently this is the only place where replicaof state can be updated on
+                     * this function, since updateShardId may update myself shard_id and caused
+                     * areInSameShard check failed. Explicitly check for a replication loop before
+                     * attempting the replication chain folding logic. */
+                    if (myself->replicaof && myself->replicaof->replicaof && myself->replicaof->replicaof != myself) {
+                        /* Safeguard against sub-replicas.
+                         *
+                         * A replica's primary can turn itself into a replica if its last slot
+                         * is removed. If no other node takes over the slot, there is nothing
+                         * else to trigger replica migration. In this case, they are not in the
+                         * same shard, so a full sync is required.
+                         *
+                         * Or a replica's primary can turn itself into a replica of its other
+                         * replica during a failover. In this case, they are in the same shard,
+                         * so we can try a psync. */
+                        serverLog(LL_NOTICE, "I'm a sub-replica! Reconfiguring myself as a replica of %.40s from %.40s",
+                                  myself->replicaof->replicaof->name, myself->replicaof->name);
+                        int are_in_same_shard = areInSameShard(myself->replicaof->replicaof, myself);
+                        clusterSetPrimary(myself->replicaof->replicaof, 1, !are_in_same_shard);
```

Comment:
the more concise single line form would be my favorite. the new variable `are_in_same_shard` doesn't provide additional context, but if you'd still prefer a variable, I would suggest `full_sync_required`
```suggestion
                        clusterSetPrimary(myself->replicaof->replicaof, 1, !areInSameShard(myself->replicaof->replicaof, myself));
```

Reply:
i am using this because the fomat action will complain it. and `are_in_same_shard` this var name, i am using it everywhere in the fisrt PR...